File: android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.java
// Copyright 2015-present 650 Industries. All rights reserved.

package host.exp.exponent;

import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import expo.modules.updates.UpdatesConfiguration;
import expo.modules.updates.UpdatesUtils;
import expo.modules.updates.db.DatabaseHolder;
import expo.modules.updates.db.entity.UpdateEntity;
import expo.modules.updates.launcher.Launcher;
import expo.modules.updates.launcher.NoDatabaseLauncher;
import expo.modules.updates.selectionpolicy.LauncherSelectionPolicyFilterAware;
import expo.modules.updates.selectionpolicy.LoaderSelectionPolicyFilterAware;
import expo.modules.updates.selectionpolicy.ReaperSelectionPolicyDevelopmentClient;
import expo.modules.updates.selectionpolicy.SelectionPolicy;
import expo.modules.updates.loader.EmbeddedLoader;
import expo.modules.updates.loader.FileDownloader;
import expo.modules.updates.loader.LoaderTask;
import expo.modules.updates.manifest.Manifest;
import expo.modules.updates.manifest.ManifestFactory;
import expo.modules.updates.manifest.raw.RawManifest;
import host.exp.exponent.di.NativeModuleDepsProvider;
import host.exp.exponent.exceptions.ManifestException;
import host.exp.exponent.kernel.ExpoViewKernel;
import host.exp.exponent.kernel.Kernel;
import host.exp.exponent.kernel.KernelConfig;
import host.exp.exponent.storage.ExponentSharedPreferences;

public class ExpoUpdatesAppLoader {

  @Inject
  ExponentManifest mExponentManifest;

  @Inject
  ExponentSharedPreferences mExponentSharedPreferences;

  @Inject
  DatabaseHolder mDatabaseHolder;

  @Inject
  Kernel mKernel;

  private static final String TAG = ExpoUpdatesAppLoader.class.getSimpleName();

  public static final String UPDATES_EVENT_NAME = "Expo.nativeUpdatesEvent";
  public static final String UPDATE_AVAILABLE_EVENT = "updateAvailable";
  public static final String UPDATE_NO_UPDATE_AVAILABLE_EVENT = "noUpdateAvailable";
  public static final String UPDATE_ERROR_EVENT = "error";

  public enum AppLoaderStatus {
    CHECKING_FOR_UPDATE, DOWNLOADING_NEW_UPDATE
  }

  private String mManifestUrl;
  private AppLoaderCallback mCallback;
  private final boolean mUseCacheOnly;

  private UpdatesConfiguration mUpdatesConfiguration;
  private File mUpdatesDirectory;
  private FileDownloader mFileDownloader;
  private SelectionPolicy mSelectionPolicy;
  private Launcher mLauncher;
  private boolean mIsEmergencyLaunch = false;
  private boolean mIsUpToDate = true;
  private AppLoaderStatus mStatus;
  private boolean mShouldShowAppLoaderStatus = true;

  private boolean isStarted = false;

  public interface AppLoaderCallback {
    void onOptimisticManifest(RawManifest optimisticManifest);
    void onManifestCompleted(RawManifest manifest);
    void onBundleCompleted(String localBundlePath);
    void emitEvent(JSONObject params);
    void updateStatus(AppLoaderStatus status);
    void onError(Exception e);
  }

  public ExpoUpdatesAppLoader(String manifestUrl, AppLoaderCallback callback) {
    this(manifestUrl, callback, false);
  }

  public ExpoUpdatesAppLoader(String manifestUrl, AppLoaderCallback callback, boolean useCacheOnly) {
    NativeModuleDepsProvider.getInstance().inject(ExpoUpdatesAppLoader.class, this);

    mManifestUrl = manifestUrl;
    mCallback = callback;
    mUseCacheOnly = useCacheOnly;
  }

  public UpdatesConfiguration getUpdatesConfiguration() {
    if (mUpdatesConfiguration == null) {
      throw new IllegalStateException("Tried to access UpdatesConfiguration before it was set");
    }
    return mUpdatesConfiguration;
  }

  public File getUpdatesDirectory() {
    if (mUpdatesDirectory == null) {
      throw new IllegalStateException("Tried to access UpdatesDirectory before it was set");
    }
    return mUpdatesDirectory;
  }

  public SelectionPolicy getSelectionPolicy() {
    if (mSelectionPolicy == null) {
      throw new IllegalStateException("Tried to access SelectionPolicy before it was set");
    }
    return mSelectionPolicy;
  }

  public FileDownloader getFileDownloader() {
    if (mFileDownloader == null) {
      throw new IllegalStateException("Tried to access FileDownloader before it was set");
    }
    return mFileDownloader;
  }

  public Launcher getLauncher() {
    if (mLauncher == null) {
      throw new IllegalStateException("Tried to access Launcher before it was set");
    }
    return mLauncher;
  }

  public boolean isEmergencyLaunch() {
    return mIsEmergencyLaunch;
  }

  public boolean isUpToDate() {
    return mIsUpToDate;
  }

  public AppLoaderStatus getStatus() {
    return mStatus;
  }

  public boolean shouldShowAppLoaderStatus() {
    return mShouldShowAppLoaderStatus;
  }

  private void updateStatus(AppLoaderStatus status) {
    mStatus = status;
    mCallback.updateStatus(status);
  }

  public void start(Context context) {
    if (isStarted) {
      throw new IllegalStateException("AppLoader for " + mManifestUrl + " was started twice. AppLoader.start() may only be called once per instance.");
    }
    isStarted = true;
    mStatus = AppLoaderStatus.CHECKING_FOR_UPDATE;

    mFileDownloader = new FileDownloader(context);
    mKernel.addAppLoaderForManifestUrl(mManifestUrl, this);

    Uri httpManifestUrl = mExponentManifest.httpManifestUrl(mManifestUrl);

    String releaseChannel = Constants.RELEASE_CHANNEL;
    if (!Constants.isStandaloneApp()) {
      // in Expo Go, the release channel can change at runtime depending on the URL we load
      String releaseChannelQueryParam = httpManifestUrl.getQueryParameter(ExponentManifest.QUERY_PARAM_KEY_RELEASE_CHANNEL);
      if (releaseChannelQueryParam != null) {
        releaseChannel = releaseChannelQueryParam;
      }
    }

    HashMap<String, Object> configMap = new HashMap<>();
    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_UPDATE_URL_KEY, httpManifestUrl);
    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_SCOPE_KEY_KEY, httpManifestUrl.toString());
    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_SDK_VERSION_KEY, Constants.SDK_VERSIONS);
    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_RELEASE_CHANNEL_KEY, releaseChannel);
    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_HAS_EMBEDDED_UPDATE_KEY, Constants.isStandaloneApp());
    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_ENABLED_KEY, Constants.ARE_REMOTE_UPDATES_ENABLED);
    if (mUseCacheOnly) {
      configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_CHECK_ON_LAUNCH_KEY, "NEVER");
      configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_LAUNCH_WAIT_MS_KEY, 0);
    } else {
      if (Constants.isStandaloneApp()) {
        configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_CHECK_ON_LAUNCH_KEY, Constants.UPDATES_CHECK_AUTOMATICALLY ? "ALWAYS" : "NEVER");
        configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_LAUNCH_WAIT_MS_KEY, Constants.UPDATES_FALLBACK_TO_CACHE_TIMEOUT);
      } else {
        configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_CHECK_ON_LAUNCH_KEY, "ALWAYS");
        configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_LAUNCH_WAIT_MS_KEY, 60000);
      }
    }

    configMap.put(UpdatesConfiguration.UPDATES_CONFIGURATION_REQUEST_HEADERS_KEY, getRequestHeaders());
    configMap.put("expectsSignedManifest", true);

    UpdatesConfiguration configuration = new UpdatesConfiguration();
    configuration.loadValuesFromMap(configMap);

    List<String> sdkVersionsList = new ArrayList<>(Constants.SDK_VERSIONS_LIST);
    sdkVersionsList.add(RNObject.UNVERSIONED);
    for (String sdkVersion : Constants.SDK_VERSIONS_LIST) {
      sdkVersionsList.add("exposdk:" + sdkVersion);
    }
    SelectionPolicy selectionPolicy = new SelectionPolicy(
            new LauncherSelectionPolicyFilterAware(sdkVersionsList),
            new LoaderSelectionPolicyFilterAware(),
            new ReaperSelectionPolicyDevelopmentClient()
    );

    File directory;
    try {
      directory = UpdatesUtils.getOrCreateUpdatesDirectory(context);
    } catch (Exception e) {
      mCallback.onError(e);
      return;
    }

    startLoaderTask(configuration, directory, selectionPolicy, context);
  }

  private void startLoaderTask(final UpdatesConfiguration configuration, final File directory, final SelectionPolicy selectionPolicy, final Context context) {
    mUpdatesConfiguration = configuration;
    mUpdatesDirectory = directory;
    mSelectionPolicy = selectionPolicy;

    if (!configuration.isEnabled()) {
      launchWithNoDatabase(context, null);
      return;
    }

    new LoaderTask(configuration, mDatabaseHolder, directory, mFileDownloader, selectionPolicy, new LoaderTask.LoaderTaskCallback() {
      private boolean didAbort = false;

      @Override
      public void onFailure(Exception e) {
        if (Constants.isStandaloneApp()) {
          mIsEmergencyLaunch = true;
          launchWithNoDatabase(context, e);
        } else {
          if (didAbort) {
            return;
          }
          Exception exception = e;
          try {
            JSONObject errorJson = new JSONObject(e.getMessage());
            exception = new ManifestException(e, mManifestUrl, errorJson);
          } catch (Exception ex) {
            // do nothing, expected if the error payload does not come from a conformant server
          }
          mCallback.onError(exception);
        }
      }

      @Override
      public boolean onCachedUpdateLoaded(UpdateEntity update) {
        setShouldShowAppLoaderStatus(update.getRawManifest());
        if (update.getRawManifest().isUsingDeveloperTool()) {
          return false;
        } else {
          try {
            // TODO(wschurman): audit ID usage for new manifests
            String experienceId = update.getRawManifest().getLegacyID();
            // if previous run of this app failed due to a loading error, we want to make sure to check for remote updates
            JSONObject experienceMetadata = mExponentSharedPreferences.getExperienceMetadata(experienceId);
            if (experienceMetadata != null && experienceMetadata.optBoolean(ExponentSharedPreferences.EXPERIENCE_METADATA_LOADING_ERROR)) {
              return false;
            }
          } catch (Exception e) {
            return true;
          }
        }
        return true;
      }

      @Override
      public void onRemoteManifestLoaded(Manifest manifest) {
        // expo-cli does not always respect our SDK version headers and respond with a compatible update or an error
        // so we need to check the compatibility here
        @Nullable String sdkVersion = manifest.getRawManifest().getSDKVersionNullable();
        if (!isValidSdkVersion(sdkVersion)) {
          mCallback.onError(formatExceptionForIncompatibleSdk(sdkVersion != null ? sdkVersion : "null"));
          didAbort = true;
          return;
        }

        setShouldShowAppLoaderStatus(manifest.getRawManifest());
        mCallback.onOptimisticManifest(manifest.getRawManifest());
        updateStatus(AppLoaderStatus.DOWNLOADING_NEW_UPDATE);
      }

      @Override
      public void onSuccess(Launcher launcher, boolean isUpToDate) {
        if (didAbort) {
          return;
        }
        mLauncher = launcher;
        mIsUpToDate = isUpToDate;
        try {
          JSONObject manifestJson = processManifestJson(launcher.getLaunchedUpdate().manifest);
          RawManifest manifest = ManifestFactory.INSTANCE.getRawManifestFromJson(manifestJson);
          mCallback.onManifestCompleted(manifest);

          // ReactAndroid will load the bundle on its own in development mode
          if (!manifest.isDevelopmentMode()) {
            mCallback.onBundleCompleted(launcher.getLaunchAssetFile());
          }
        } catch (Exception e) {
          mCallback.onError(e);
        }
      }

      @Override
      public void onBackgroundUpdateFinished(LoaderTask.BackgroundUpdateStatus status, @Nullable UpdateEntity update, @Nullable Exception exception) {
        if (didAbort) {
          return;
        }
        try {
          JSONObject jsonParams = new JSONObject();
          if (status == LoaderTask.BackgroundUpdateStatus.ERROR) {
            if (exception == null) {
              throw new AssertionError("Background update with error status must have a nonnull exception object");
            }
            jsonParams.put("type", UPDATE_ERROR_EVENT);
            jsonParams.put("message", exception.getMessage());
          } else if (status == LoaderTask.BackgroundUpdateStatus.UPDATE_AVAILABLE) {
            if (update == null) {
              throw new AssertionError("Background update with error status must have a nonnull update object");
            }
            jsonParams.put("type", UPDATE_AVAILABLE_EVENT);
            jsonParams.put("manifestString", update.manifest.toString());
          } else if (status == LoaderTask.BackgroundUpdateStatus.NO_UPDATE_AVAILABLE) {
            jsonParams.put("type", UPDATE_NO_UPDATE_AVAILABLE_EVENT);
          }
          mCallback.emitEvent(jsonParams);
        } catch (Exception e) {
          Log.e(TAG, "Failed to emit event to JS", e);
        }
      }
    }).start(context);
  }

  private void launchWithNoDatabase(Context context, Exception e) {
    mLauncher = new NoDatabaseLauncher(context, mUpdatesConfiguration, e);

    JSONObject manifestJson = EmbeddedLoader.readEmbeddedManifest(context, mUpdatesConfiguration).getRawManifest().getRawJson();
    try {
      manifestJson = processManifestJson(manifestJson);
    } catch (Exception ex) {
      Log.e(TAG, "Failed to process manifest; attempting to launch with raw manifest. This may cause errors or unexpected behavior.", e);
    }
    mCallback.onManifestCompleted(ManifestFactory.INSTANCE.getRawManifestFromJson(manifestJson));

    String launchAssetFile = mLauncher.getLaunchAssetFile();
    if (launchAssetFile == null) {
      // ReactInstanceManagerBuilder accepts embedded assets as strings with "assets://" prefixed
      launchAssetFile = "assets://" + mLauncher.getBundleAssetName();
    }
    mCallback.onBundleCompleted(launchAssetFile);
  }

  private JSONObject processManifestJson(JSONObject manifestJson) throws JSONException {
    Uri parsedManifestUrl = Uri.parse(mManifestUrl);
    if (!manifestJson.optBoolean(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, false) &&
        isThirdPartyHosted(parsedManifestUrl) &&
        !Constants.isStandaloneApp()) {
      // Sandbox third party apps and consider them verified
      // for https urls, sandboxed id is of form quinlanj.github.io/myProj-myApp
      // for http urls, sandboxed id is of form UNVERIFIED-quinlanj.github.io/myProj-myApp
      String protocol = parsedManifestUrl.getScheme();
      String securityPrefix = protocol.equals("https") || protocol.equals("exps") ? "" : "UNVERIFIED-";
      String path = parsedManifestUrl.getPath() != null ? parsedManifestUrl.getPath() : "";
      String slug = manifestJson.has(ExponentManifest.MANIFEST_SLUG) ? manifestJson.getString(ExponentManifest.MANIFEST_SLUG) : "";
      String sandboxedId = securityPrefix + parsedManifestUrl.getHost() + path + "-" + slug;
      manifestJson.put(ExponentManifest.MANIFEST_ID_KEY, sandboxedId);
      manifestJson.put(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, true);
    }
    if (Constants.isStandaloneApp()) {
      manifestJson.put(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, true);
    }
    if (!manifestJson.has(ExponentManifest.MANIFEST_IS_VERIFIED_KEY)) {
      manifestJson.put(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, false);
    }

    if (!manifestJson.optBoolean(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, false) &&
        mExponentManifest.isAnonymousExperience(ManifestFactory.INSTANCE.getRawManifestFromJson(manifestJson))) {
      // automatically verified
      manifestJson.put(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, true);
    }

    return manifestJson;
  }

  private boolean isThirdPartyHosted(Uri uri) {
    String host = uri.getHost();
    return !(host.equals("exp.host") || host.equals("expo.io") || host.equals("exp.direct") || host.equals("expo.test") ||
      host.endsWith(".exp.host") || host.endsWith(".expo.io") || host.endsWith(".exp.direct") || host.endsWith(".expo.test"));
  }

  private void setShouldShowAppLoaderStatus(RawManifest manifest) {
    // we don't want to show the cached experience alert when Updates.reloadAsync() is called
    if (mUseCacheOnly) {
      mShouldShowAppLoaderStatus = false;
      return;
    }

    mShouldShowAppLoaderStatus = !manifest.isDevelopmentSilentLaunch();
  }

  private Map<String, String> getRequestHeaders() {
    HashMap<String, String> headers = new HashMap<>();
    headers.put("Expo-Updates-Environment", getClientEnvironment());
    headers.put("Expo-Client-Environment", getClientEnvironment());

    if (ExpoViewKernel.getInstance().getVersionName() != null) {
      headers.put("Exponent-Version", ExpoViewKernel.getInstance().getVersionName());
    }

    String sessionSecret = mExponentSharedPreferences.getSessionSecret();
    if (sessionSecret != null) {
      headers.put("Expo-Session", sessionSecret);
    }

    // XDL expects the full "exponent-" header names
    headers.put("Exponent-Accept-Signature", "true");
    headers.put("Exponent-Platform", "android");
    if (KernelConfig.FORCE_UNVERSIONED_PUBLISHED_EXPERIENCES) {
      headers.put("Exponent-SDK-Version", "UNVERSIONED");
    } else {
      headers.put("Exponent-SDK-Version", Constants.SDK_VERSIONS);
    }

    return headers;
  }

  private String getClientEnvironment() {
    if (Constants.isStandaloneApp()) {
      return "STANDALONE";
    } else if (Build.FINGERPRINT.contains("vbox") || Build.FINGERPRINT.contains("generic")) {
      return "EXPO_SIMULATOR";
    } else {
      return "EXPO_DEVICE";
    }
  }

  private boolean isValidSdkVersion(@Nullable String sdkVersion) {
    if (sdkVersion == null) {
      return false;
    }

    if (RNObject.UNVERSIONED.equals(sdkVersion)) {
      return true;
    }

    for (final String version : Constants.SDK_VERSIONS_LIST) {
      if (version.equals(sdkVersion)) {
        return true;
      }
    }

    return false;
  }

  private ManifestException formatExceptionForIncompatibleSdk(String sdkVersion) {
    JSONObject errorJson = new JSONObject();
    try {
      errorJson.put("message", "Invalid SDK version");
      if (ABIVersion.toNumber(sdkVersion) > ABIVersion.toNumber(Constants.SDK_VERSIONS_LIST.get(0))) {
        errorJson.put("errorCode", "EXPERIENCE_SDK_VERSION_TOO_NEW");
      } else {
        errorJson.put("errorCode", "EXPERIENCE_SDK_VERSION_OUTDATED");
        errorJson.put("metadata", new JSONObject().put(
          "availableSDKVersions",
          new JSONArray().put(sdkVersion))
        );
      }
    } catch (Exception e) {
      Log.e(TAG, "Failed to format error message for incompatible SDK version", e);
    }
    return new ManifestException(new Exception("Incompatible SDK version"), mManifestUrl, errorJson);
  }
}

File: android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package host.exp.exponent

import android.content.Context
import android.net.Uri
import android.util.Log
import expo.modules.jsonutils.getNullable
import expo.modules.manifests.core.LegacyManifest
import expo.modules.core.utilities.EmulatorUtilities
import expo.modules.updates.UpdatesConfiguration
import expo.modules.updates.UpdatesUtils
import expo.modules.updates.db.DatabaseHolder
import expo.modules.updates.db.entity.UpdateEntity
import expo.modules.updates.launcher.Launcher
import expo.modules.updates.loader.FileDownloader
import expo.modules.updates.loader.LoaderTask
import expo.modules.updates.loader.LoaderTask.LoaderTaskCallback
import expo.modules.updates.loader.LoaderTask.RemoteUpdateStatus
import expo.modules.updates.manifest.UpdateManifest
import expo.modules.manifests.core.Manifest
import expo.modules.updates.codesigning.CODE_SIGNING_METADATA_ALGORITHM_KEY
import expo.modules.updates.codesigning.CODE_SIGNING_METADATA_KEY_ID_KEY
import expo.modules.updates.codesigning.CodeSigningAlgorithm
import expo.modules.updates.selectionpolicy.LoaderSelectionPolicyFilterAware
import expo.modules.updates.selectionpolicy.ReaperSelectionPolicyDevelopmentClient
import expo.modules.updates.selectionpolicy.SelectionPolicy
import host.exp.exponent.di.NativeModuleDepsProvider
import host.exp.exponent.exceptions.ManifestException
import host.exp.exponent.kernel.ExperienceKey
import host.exp.exponent.kernel.ExpoViewKernel
import host.exp.exponent.kernel.Kernel
import host.exp.exponent.kernel.KernelConfig
import host.exp.exponent.storage.ExponentSharedPreferences
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.util.*
import javax.inject.Inject

private const val UPDATE_AVAILABLE_EVENT = "updateAvailable"
private const val UPDATE_NO_UPDATE_AVAILABLE_EVENT = "noUpdateAvailable"
private const val UPDATE_ERROR_EVENT = "error"

/**
 * Entry point to expo-updates in Expo Go. Fulfills many of the
 * purposes of [UpdatesController] along with serving as an interface to the rest of the kernel.
 *
 * Dynamically generates a configuration object with the correct scope key, and then, like
 * [UpdatesController], delegates to an instance of [LoaderTask] to start the process of loading and
 * launching an update, and responds appropriately depending on the callbacks that are invoked.
 *
 * Multiple instances of ExpoUpdatesAppLoader can exist at a time; instances are retained by
 * [Kernel].
 */
class ExpoUpdatesAppLoader @JvmOverloads constructor(
  private val manifestUrl: String,
  private val callback: AppLoaderCallback,
  private val useCacheOnly: Boolean = false
) {
  @Inject
  lateinit var exponentManifest: ExponentManifest

  @Inject
  lateinit var exponentSharedPreferences: ExponentSharedPreferences

  @Inject
  lateinit var databaseHolder: DatabaseHolder

  @Inject
  lateinit var kernel: Kernel

  enum class AppLoaderStatus {
    CHECKING_FOR_UPDATE, DOWNLOADING_NEW_UPDATE
  }

  var isEmergencyLaunch = false
    private set
  var isUpToDate = true
    private set
  var status: AppLoaderStatus? = null
    private set
  var shouldShowAppLoaderStatus = true
    private set
  private var isStarted = false

  interface AppLoaderCallback {
    fun onOptimisticManifest(optimisticManifest: Manifest)
    fun onManifestCompleted(manifest: Manifest)
    fun onBundleCompleted(localBundlePath: String)
    fun emitEvent(params: JSONObject)
    fun updateStatus(status: AppLoaderStatus)
    fun onError(e: Exception)
  }

  lateinit var updatesConfiguration: UpdatesConfiguration
    private set

  lateinit var updatesDirectory: File
    private set

  lateinit var selectionPolicy: SelectionPolicy
    private set

  lateinit var fileDownloader: FileDownloader
    private set

  lateinit var launcher: Launcher
    private set

  private fun updateStatus(status: AppLoaderStatus) {
    this.status = status
    callback.updateStatus(status)
  }

  fun start(context: Context) {
    check(!isStarted) { "AppLoader for $manifestUrl was started twice. AppLoader.start() may only be called once per instance." }
    isStarted = true
    status = AppLoaderStatus.CHECKING_FOR_UPDATE
    fileDownloader = FileDownloader(context)
    kernel.addAppLoaderForManifestUrl(manifestUrl, this)
    val httpManifestUrl = exponentManifest.httpManifestUrl(manifestUrl)
    val configMap = mutableMapOf<String, Any>()
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_UPDATE_URL_KEY] = httpManifestUrl
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_SCOPE_KEY_KEY] = httpManifestUrl.toString()
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_SDK_VERSION_KEY] = Constants.SDK_VERSIONS
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_HAS_EMBEDDED_UPDATE_KEY] = false
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_ENABLED_KEY] = true
    if (useCacheOnly) {
      configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_CHECK_ON_LAUNCH_KEY] = "NEVER"
      configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_LAUNCH_WAIT_MS_KEY] = 0
    } else {
      configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_CHECK_ON_LAUNCH_KEY] = "ALWAYS"
      configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_LAUNCH_WAIT_MS_KEY] = 60000
    }
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_REQUEST_HEADERS_KEY] = requestHeaders
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_EXPECTS_EXPO_SIGNED_MANIFEST] = true
    // in Expo Go, embed the Expo Root Certificate and get the Expo Go intermediate certificate and development certificates from the multipart manifest response part
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_CODE_SIGNING_CERTIFICATE] = context.assets.open("expo-root.pem").readBytes().decodeToString()
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_CODE_SIGNING_METADATA] = mapOf(
      CODE_SIGNING_METADATA_KEY_ID_KEY to "expo-root",
      CODE_SIGNING_METADATA_ALGORITHM_KEY to CodeSigningAlgorithm.RSA_SHA256.algorithmName,
    )
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_CODE_SIGNING_INCLUDE_MANIFEST_RESPONSE_CERTIFICATE_CHAIN] = true
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_CODE_SIGNING_ALLOW_UNSIGNED_MANIFESTS] = true
    // in Expo Go, ignore directives in manifest responses and require a manifest. the current directives (no update available, roll back) don't have any practical use outside of apps that use the library directly.
    configMap[UpdatesConfiguration.UPDATES_CONFIGURATION_ENABLE_EXPO_UPDATES_PROTOCOL_V0_COMPATIBILITY_MODE] = true

    val configuration = UpdatesConfiguration(null, configMap)
    val sdkVersionsList = mutableListOf<String>().apply {
      (Constants.SDK_VERSIONS_LIST + listOf(RNObject.UNVERSIONED)).forEach {
        add(it)
        add("exposdk:$it")
      }
    }
    val selectionPolicy = SelectionPolicy(
      ExpoGoLauncherSelectionPolicyFilterAware(sdkVersionsList),
      LoaderSelectionPolicyFilterAware(),
      ReaperSelectionPolicyDevelopmentClient()
    )
    val directory: File = try {
      UpdatesUtils.getOrCreateUpdatesDirectory(context)
    } catch (e: Exception) {
      callback.onError(e)
      return
    }
    startLoaderTask(configuration, directory, selectionPolicy, context)
  }

  private fun startLoaderTask(
    configuration: UpdatesConfiguration,
    directory: File,
    selectionPolicy: SelectionPolicy,
    context: Context
  ) {
    updatesConfiguration = configuration
    updatesDirectory = directory
    this.selectionPolicy = selectionPolicy
    LoaderTask(
      configuration,
      databaseHolder,
      directory,
      fileDownloader,
      selectionPolicy,
      object : LoaderTaskCallback {
        private var didAbort = false
        override fun onFailure(e: Exception) {
          if (didAbort) {
            return
          }
          var exception = e
          try {
            val errorJson = JSONObject(e.message!!)
            exception = ManifestException(e, manifestUrl, errorJson)
          } catch (ex: Exception) {
            // do nothing, expected if the error payload does not come from a conformant server
          }
          callback.onError(exception)
        }

        override fun onCachedUpdateLoaded(update: UpdateEntity): Boolean {
          val manifest = Manifest.fromManifestJson(update.manifest)
          setShouldShowAppLoaderStatus(manifest)
          if (manifest.isUsingDeveloperTool()) {
            return false
          } else {
            try {
              val experienceKey = ExperienceKey.fromManifest(manifest)
              // if previous run of this app failed due to a loading error, we want to make sure to check for remote updates
              val experienceMetadata = exponentSharedPreferences.getExperienceMetadata(experienceKey)
              if (experienceMetadata != null && experienceMetadata.optBoolean(
                  ExponentSharedPreferences.EXPERIENCE_METADATA_LOADING_ERROR
                )
              ) {
                return false
              }
            } catch (e: Exception) {
              return true
            }
          }
          return true
        }

        override fun onRemoteUpdateManifestResponseManifestLoaded(updateManifest: UpdateManifest) {
          // expo-cli does not always respect our SDK version headers and respond with a compatible update or an error
          // so we need to check the compatibility here
          val sdkVersion = updateManifest.manifest.getExpoGoSDKVersion()
          if (!isValidSdkVersion(sdkVersion)) {
            callback.onError(formatExceptionForIncompatibleSdk(sdkVersion))
            didAbort = true
            return
          }
          setShouldShowAppLoaderStatus(updateManifest.manifest)
          callback.onOptimisticManifest(updateManifest.manifest)
          updateStatus(AppLoaderStatus.DOWNLOADING_NEW_UPDATE)
        }

        override fun onSuccess(launcher: Launcher, isUpToDate: Boolean) {
          if (didAbort) {
            return
          }
          this@ExpoUpdatesAppLoader.launcher = launcher
          this@ExpoUpdatesAppLoader.isUpToDate = isUpToDate
          try {
            val manifestJson = processManifestJson(launcher.launchedUpdate!!.manifest)
            val manifest = Manifest.fromManifestJson(manifestJson)
            callback.onManifestCompleted(manifest)

            // ReactAndroid will load the bundle on its own in development mode
            if (!manifest.isDevelopmentMode()) {
              callback.onBundleCompleted(launcher.launchAssetFile!!)
            }
          } catch (e: Exception) {
            callback.onError(e)
          }
        }

        override fun onRemoteUpdateFinished(
          status: RemoteUpdateStatus,
          update: UpdateEntity?,
          exception: Exception?
        ) {
          if (didAbort) {
            return
          }
          try {
            val jsonParams = JSONObject()
            when (status) {
              RemoteUpdateStatus.ERROR -> {
                if (exception == null) {
                  throw AssertionError("Background update with error status must have a nonnull exception object")
                }
                jsonParams.put("type", UPDATE_ERROR_EVENT)
                jsonParams.put("message", exception.message)
              }
              RemoteUpdateStatus.UPDATE_AVAILABLE -> {
                if (update == null) {
                  throw AssertionError("Background update with error status must have a nonnull update object")
                }
                jsonParams.put("type", UPDATE_AVAILABLE_EVENT)
                jsonParams.put("manifestString", update.manifest.toString())
              }
              RemoteUpdateStatus.NO_UPDATE_AVAILABLE -> {
                jsonParams.put("type", UPDATE_NO_UPDATE_AVAILABLE_EVENT)
              }
            }
            callback.emitEvent(jsonParams)
          } catch (e: Exception) {
            Log.e(TAG, "Failed to emit event to JS", e)
          }
        }
      }
    ).start(context)
  }

  @Throws(JSONException::class)
  private fun processManifestJson(manifestJson: JSONObject): JSONObject {
    val parsedManifestUrl = Uri.parse(manifestUrl)

    // If legacy manifest is not yet verified, served by a third party, and not an anonymous experience
    // then scope it locally by using the manifest URL as a scopeKey (id) and consider it verified.
    if (!manifestJson.optBoolean(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, false) &&
      isThirdPartyHosted(parsedManifestUrl) &&
      !exponentManifest.isAnonymousExperience(Manifest.fromManifestJson(manifestJson)) &&
      Manifest.fromManifestJson(manifestJson) is LegacyManifest
    ) {
      // for https urls, sandboxed id is of form quinlanj.github.io/myProj-myApp
      // for http urls, sandboxed id is of form UNVERIFIED-quinlanj.github.io/myProj-myApp
      val protocol = parsedManifestUrl.scheme
      val securityPrefix = if (protocol == "https" || protocol == "exps") "" else "UNVERIFIED-"
      val path = if (parsedManifestUrl.path != null) parsedManifestUrl.path else ""
      val slug = manifestJson.getNullable<String>(ExponentManifest.MANIFEST_SLUG) ?: ""
      val sandboxedId = securityPrefix + parsedManifestUrl.host + path + "-" + slug
      manifestJson.put(ExponentManifest.MANIFEST_ID_KEY, sandboxedId)
      manifestJson.put(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, true)
    }

    // if the manifest is scoped to a random anonymous scope key, automatically verify it
    if (exponentManifest.isAnonymousExperience(Manifest.fromManifestJson(manifestJson))) {
      manifestJson.put(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, true)
    }

    // otherwise set verified to false
    if (!manifestJson.has(ExponentManifest.MANIFEST_IS_VERIFIED_KEY)) {
      manifestJson.put(ExponentManifest.MANIFEST_IS_VERIFIED_KEY, false)
    }

    return manifestJson
  }

  private fun isThirdPartyHosted(uri: Uri): Boolean {
    val host = uri.host
    return !(
      host == "exp.host" || host == "expo.io" || host == "exp.direct" || host == "expo.test" ||
        host!!.endsWith(".exp.host") || host.endsWith(".expo.io") || host.endsWith(".exp.direct") || host.endsWith(
        ".expo.test"
      )
      )
  }

  private fun setShouldShowAppLoaderStatus(manifest: Manifest) {
    // we don't want to show the cached experience alert when Updates.reloadAsync() is called
    if (useCacheOnly) {
      shouldShowAppLoaderStatus = false
      return
    }
    shouldShowAppLoaderStatus = !manifest.isDevelopmentSilentLaunch()
  }

  // XDL expects the full "exponent-" header names
  private val requestHeaders: Map<String, String?>
    get() {
      val headers = mutableMapOf<String, String>()
      headers["Expo-Updates-Environment"] = clientEnvironment
      headers["Expo-Client-Environment"] = clientEnvironment
      val versionName = ExpoViewKernel.instance.versionName
      if (versionName != null) {
        headers["Exponent-Version"] = versionName
      }
      val sessionSecret = exponentSharedPreferences.sessionSecret
      if (sessionSecret != null) {
        headers["Expo-Session"] = sessionSecret
      }

      // XDL expects the full "exponent-" header names
      headers["Exponent-Accept-Signature"] = "true"
      headers["Exponent-Platform"] = "android"
      if (KernelConfig.FORCE_UNVERSIONED_PUBLISHED_EXPERIENCES) {
        headers["Exponent-SDK-Version"] = "UNVERSIONED"
      } else {
        headers["Exponent-SDK-Version"] = Constants.SDK_VERSIONS
      }
      return headers
    }

  private val isRunningOnEmulator: Boolean
    get() = EmulatorUtilities.isRunningOnEmulator()

  private val clientEnvironment: String
    get() = if (EmulatorUtilities.isRunningOnEmulator()) {
      "EXPO_SIMULATOR"
    } else {
      "EXPO_DEVICE"
    }

  private fun isValidSdkVersion(sdkVersion: String?): Boolean {
    if (sdkVersion == null) {
      return false
    }
    if (RNObject.UNVERSIONED == sdkVersion) {
      return true
    }
    for (version in Constants.SDK_VERSIONS_LIST) {
      if (version == sdkVersion) {
        return true
      }
    }
    return false
  }

  private fun formatExceptionForIncompatibleSdk(sdkVersion: String?): ManifestException {
    val errorJson = JSONObject()
    try {
      errorJson.put("message", "Invalid SDK version")
      if (sdkVersion == null) {
        errorJson.put("errorCode", "NO_SDK_VERSION_SPECIFIED")
      } else if (ABIVersion.toNumber(sdkVersion) > ABIVersion.toNumber(Constants.SDK_VERSIONS_LIST[0])) {
        errorJson.put("errorCode", "EXPERIENCE_SDK_VERSION_TOO_NEW")
      } else {
        errorJson.put("errorCode", "EXPERIENCE_SDK_VERSION_OUTDATED")
        errorJson.put(
          "metadata",
          JSONObject().put(
            "availableSDKVersions",
            JSONArray().put(sdkVersion)
          )
        )
      }
    } catch (e: Exception) {
      Log.e(TAG, "Failed to format error message for incompatible SDK version", e)
    }
    return ManifestException(Exception("Incompatible SDK version"), manifestUrl, errorJson)
  }

  companion object {
    private val TAG = ExpoUpdatesAppLoader::class.java.simpleName
    const val UPDATES_EVENT_NAME = "Expo.nativeUpdatesEvent"
  }

  init {
    NativeModuleDepsProvider.instance.inject(ExpoUpdatesAppLoader::class.java, this)
  }
}
